File: Algorithm/Data Structure/219. Contains Duplicate II.py
# 50
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        dic = {}
        for i, x in enumerate(nums):
            if x in dic:
                if i - dic[x] <= k:
                    return True
            dic[x] = i
        return False 
            

        # 시간초과 
        # s = set()
        # for i in range(len(nums)):
        #     print(s)
        #     if nums[i] in s:
        #         return True
        #     s.add(nums[i])
        #     if i >= k:
        #         s.remove(nums[i-k])
        # return False
    

File: Algorithm/Data Structure/219. Contains Duplicate II.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        if(k<=0){return false;}
        Map<Integer, Integer> last = new HashMap<>();
        for(int i=0;i<nums.length;i++){
            int x = nums[i];
            Integer l = last.get(x);
            if(l != null && i - l <= k){
                return true;
            }
            last.put(x, i);
        }
        return false;
    }
}
